File: R/sylbreak4all.R
segment_line <- function(line, pattern, separator) {
  line <- gsub("\\s+", "", line, perl = TRUE)
  gsub(pattern, paste0(separator, "\\1"), line, perl = TRUE)
}

patterns <- list(
  bm = "((?<!္)[က-အ](?![်္])|[a-zA-Z0-9ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@\\[-`’“”{-~\\s…])",
  rk = "((?<!္)[က-အ](?![်္])|[a-zA-Z0-9ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@\\[-`’“”{-~\\s…])",
  dw = "((?<!္)[က-အ](?![်္])|[a-zA-Z0-9ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@\\[-`’“”{-~\\s…])",
  bk = "((?<!္)[က-အ](?![်္])|[a-zA-Z0-9ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@\\[-`’“”{-~\\s…])",
  po = "((?<!္)[က-အ](?![်္])|[a-zA-Z0-9ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@\\[-`’“”{-~\\s…])",
  sh = "([ၵၶငၸသၺတထၼပၽၾမယရလဝႁဢၹၷႀၻၿ](?![်])|[a-zA-Z0-9႟႐-႙၊။!-/:-@\\[-`{-~\\s])",
  sk = "([ကခဂဃငစဆဇညတထဒနပဖဘမယရလဝသဟအဧ]|[a-zA-Z0-9ဒမၡဧ၀-၉၊။!-/:-@\\[(-`{-~\\s])",
  pk = "([က-အ]|[a-zA-Z0-9ၥၦၡဧ၀-၉၊။!-/:-@\\[(-`{-~\\s])",
  mo = "((?<!္)[ကခဂဃၚစဆဇၛဉညဋဌဍဎဏတထဒဓနပဖဗဘမယရလဝသဟဠၜအၝ](?![်္])|[a-zA-Z0-9ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@\\[-`’“”{-~\\s…])"
)

HELP_TEXT <- paste0(
  "Syllable Breaker Tool\n",
  "Version: 0.9.0\n\n",
  "Supported Languages:\n",
  "    bm  - Burmese\n",
  "    bk  - Beik\n",
  "    dw  - Dawei\n",
  "    rk  - Rakhine\n",
  "    mo  - Mon\n",
  "    po  - PaO\n",
  "    sh  - Shan\n",
  "    sk  - Sgaw Kayin\n",
  "    pk  - Pwo Kayin\n\n",
  "Usage: sylbreak4all [-i filename] [-l language] [-s separator] [-p]\n",
  "    -i | --input-file   Input file name (default: STDIN)\n",
  "    -s | --separator    Separator (default: '|')\n",
  "    -l | --language     Language (default: bm for Burmese)\n",
  "    -p | --print        Print input alongside the output (default: 0)\n",
  "    -h | --help         Show this help\n"
)

args <- commandArgs(trailingOnly = TRUE)
input_file <- NULL; language <- "bm"; separator <- "|"; print_input <- FALSE
idx <- 1
while (idx <= length(args)) {
  arg <- args[[idx]]
  if (arg %in% c("-h", "--help")) { cat(HELP_TEXT); quit(status = 0)
  } else if (arg %in% c("-i", "--input", "--input-file")) {
    idx <- idx + 1; if (idx > length(args)) stop("Missing value for -i/--input-file"); input_file <- args[[idx]]
  } else if (arg %in% c("-l", "--language", "--lang")) {
    idx <- idx + 1; if (idx > length(args)) stop("Missing value for -l/--language"); language <- args[[idx]]
  } else if (arg %in% c("-s", "--separator")) {
    idx <- idx + 1; if (idx > length(args)) stop("Missing value for -s/--separator"); separator <- args[[idx]]
  } else if (arg %in% c("-p", "--print")) { print_input <- TRUE
  } else { stop(paste("Unknown argument:", arg)) }
  idx <- idx + 1
}
if (!(language %in% names(patterns))) stop(paste("Unsupported language:", language))

con <- if (is.null(input_file)) file("stdin", "r", encoding = "UTF-8") else file(input_file, "r", encoding = "UTF-8")
while (TRUE) {
  line <- readLines(con, n = 1, warn = FALSE)
  if (length(line) == 0) break
  clean <- gsub("\\s+", "", line, perl = TRUE)
  output <- segment_line(line, patterns[[language]], separator)
  if (print_input) { cat("input: ", clean, "\n", sep = ""); cat("output: ", output, "\n", sep = "") }
  else cat(output, "\n", sep = "")
}
close(con)

File: python/sylbreak4all.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""sylbreak4all: regex-based syllable breaking for nine Myanmar languages."""

from __future__ import annotations

import re
import sys
from pathlib import Path
from typing import TextIO

MY_CONSONANT = r"က-အ"
EN_CHAR = r"a-zA-Z0-9"
SS_SYMBOL = "္"
ATHAT = "်"
OTHER_CHAR = r"ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@[-`’“”{-~\s…"

SH_CONSONANT = r"ၵၶငၸသၺတထၼပၽၾမယရလဝႁဢၹၷႀၻၿ"
OTHER_SH = r"႟႐-႙၊။!-/:-@[-`{-~\s"
SK_CONSONANT = r"ကခဂဃငစဆဇညတထဒနပဖဘမယရလဝသဟအဧ"
OTHER_SK = r"ဒမၡဧ၀-၉၊။!-/:-@[( -`{-~\s".replace(" ", "")
OTHER_PK = r"ၥၦၡဧ၀-၉၊။!-/:-@[( -`{-~\s".replace(" ", "")
MO_CONSONANT = r"ကခဂဃၚစဆဇၛဉညဋဌဍဎဏတထဒဓနပဖဗဘမယရလဝသဟဠၜအၝ"


def compile_patterns() -> dict[str, re.Pattern[str]]:
    common = rf"((?<!{SS_SYMBOL})[{MY_CONSONANT}](?![{ATHAT}{SS_SYMBOL}])|[{EN_CHAR}{OTHER_CHAR}])"
    return {
        "bm": re.compile(common),
        "rk": re.compile(common),
        "dw": re.compile(common),
        "bk": re.compile(common),
        "po": re.compile(common),
        "sh": re.compile(rf"([{SH_CONSONANT}](?![{ATHAT}])|[{EN_CHAR}{OTHER_SH}])"),
        "sk": re.compile(rf"([{SK_CONSONANT}]|[{EN_CHAR}{OTHER_SK}])"),
        "pk": re.compile(rf"([{MY_CONSONANT}]|[{EN_CHAR}{OTHER_PK}])"),
        "mo": re.compile(rf"((?<!{SS_SYMBOL})[{MO_CONSONANT}](?![{ATHAT}{SS_SYMBOL}])|[{EN_CHAR}{OTHER_CHAR}])"),
    }


def segment_line(line: str, pattern: re.Pattern[str], separator: str) -> str:
    # Match the historical sylbreak4all reference behavior: phrase whitespace is
    # removed before applying the syllable-breaking RE.
    line = re.sub(r"\s+", "", line)
    output = pattern.sub(lambda m: separator + m.group(1), line)
    return re.sub(r"^\s+|\s+$", "", output)


def open_input(path: str | None) -> TextIO:
    if path is None:
        return sys.stdin
    return Path(path).open("r", encoding="utf-8")


HELP_TEXT = """Syllable Breaker Tool
Version: 0.9.0

Supported Languages:
    bm  - Burmese
    bk  - Beik
    dw  - Dawei
    rk  - Rakhine
    mo  - Mon
    po  - PaO
    sh  - Shan
    sk  - Sgaw Kayin
    pk  - Pwo Kayin

Usage: sylbreak4all [-i filename] [-l language] [-s separator] [-p]
    -i | --input-file   Input file name (default: STDIN)
    -s | --separator    Separator (default: '|')
    -l | --language     Language (default: bm for Burmese)
    -p | --print        Print input alongside the output (default: 0)
    -h | --help         Show this help
"""

def print_help(stream: TextIO = sys.stdout) -> None:
    print(HELP_TEXT, end="", file=stream)

def main() -> int:
    input_file = None
    separator = "|"
    language = "bm"
    print_input = False
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ("-h", "--help"):
            print_help()
            return 0
        if arg in ("-i", "--input", "--input-file"):
            i += 1
            if i >= len(args):
                print("Missing value for -i/--input-file", file=sys.stderr)
                print_help(sys.stderr)
                return 1
            input_file = args[i]
        elif arg in ("-s", "--separator"):
            i += 1
            if i >= len(args):
                print("Missing value for -s/--separator", file=sys.stderr)
                print_help(sys.stderr)
                return 1
            separator = args[i]
        elif arg in ("-l", "--language", "--lang"):
            i += 1
            if i >= len(args):
                print("Missing value for -l/--language", file=sys.stderr)
                print_help(sys.stderr)
                return 1
            language = args[i]
        elif arg in ("-p", "--print"):
            print_input = True
        else:
            print(f"Unknown argument: {arg}", file=sys.stderr)
            print_help(sys.stderr)
            return 1
        i += 1

    patterns = compile_patterns()
    if language not in patterns:
        print(f"Unsupported language: {language}", file=sys.stderr)
        print_help(sys.stderr)
        return 1

    fh = open_input(input_file)
    try:
        for raw_line in fh:
            line = raw_line.rstrip("\r\n")
            output = segment_line(line, patterns[language], separator)
            if print_input:
                clean = re.sub(r"\s+", "", line)
                print(f"input: {clean}")
                print(f"output: {output}")
            else:
                print(output)
    finally:
        if fh is not sys.stdin:
            fh.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
